neptune/observation.go

```diff
@@ -2,15 +2,109 @@ package neptune
 
 import (
 	"context"
+	"fmt"
+	"strings"
+	"time"
 
+	"github.com/pkg/errors"
+
+	"github.com/ONSdigital/dp-graph/neptune/query"
 	"github.com/ONSdigital/dp-graph/observation"
 	"github.com/ONSdigital/dp-observation-importer/models"
+	"github.com/ONSdigital/go-ns/log"
 )
 
+// ErrInvalidFilter is returned if the provided filter is nil.
+var ErrInvalidFilter = errors.New("nil filter cannot be processed")
+
 func (n *NeptuneDB) StreamCSVRows(ctx context.Context, filter *observation.Filter, limit *int) (observation.StreamRowReader, error) {
-	return nil, nil
+	if filter == nil {
+		return nil, ErrInvalidFilter
+	}
+
+	q := fmt.Sprintf(query.GetInstanceHeaderPart, filter.InstanceID)
+
+	q += buildObservationsQuery(filter)
+	q += query.GetObservationSelectRowPart
+
+	if limit != nil {
+		q += fmt.Sprintf(query.LimitPart, *limit)
+	}
+
+	return n.Pool.OpenStreamCursor(ctx, q, nil, nil)
+}
+
+func buildObservationsQuery(f *observation.Filter) string {
+	if f.IsEmpty() {
+		return fmt.Sprintf(query.GetAllObservationsPart, f.InstanceID)
+	}
+
+	q := fmt.Sprintf(query.GetObservationsPart, f.InstanceID)
+
+	for _, dim := range f.DimensionFilters {
+		if len(dim.Options) == 0 {
+			continue
+		}
+
+		for i, opt := range dim.Options {
+			dim.Options[i] = fmt.Sprintf("'%s'", opt)
+		}
+
+		q += fmt.Sprintf(query.GetObservationDimensionPart, f.InstanceID, dim.Name, strings.Join(dim.Options, ",")) + ","
+	}
+
+	//remove trailing comma and close match statement
+	q = strings.Trim(q, ",")
+	q += ")"
+
+	return q
 }
 
+// TODO: this global state is only used for metrics, not actual code flow,
+// but should be revisited before production use
+var batchCount = 0
+var totalTime time.Time
```

Comment:
I think by convention all package vars should be declared at the top of the file - they are easily lost and hard to find if they are scattered inbetween methods.